python/cucumber_expressions/tree_regexp.py

```diff
@@ -51,22 +57,30 @@ def create_group_builder(self, regexp):
         return stack.pop()
 
     @staticmethod
-    def is_non_capturing(source, index):
-        # Regex is valid. Bounds check not required.
+    def is_named_group(source: str, index: int) -> bool:
+        """
+        Check if the group at the given index is a named capturing group, e.g. (?P<name>...).
+        """
+        return source[index + 1 : index + 3] == "P<" and source[index + 3] != "?"
+
+    @staticmethod
+    def extract_named_group_name(source: str, index: int) -> str:
+        """
+        Extract the name of a named capturing group, e.g., (?P<name>...) returns "name".
+        """
+        group_name_start = index + 3
+        group_name_end = source.find(">", group_name_start)
+        return source[group_name_start:group_name_end]
+
+    @staticmethod
+    def is_non_capturing(source: str, index: int) -> bool:
+        # Check if it's a non-capturing group like (?:...)
         if source[index + 1] != "?":
-            # (X)
             return False
-        if source[index + 2] != "<":
-            # (?:X)
-            # (?idmsuxU-idmsuxU)
-            # (?idmsux-idmsux:X)
-            # (?=X)
-            # (?!X)
-            # (?>X)
-            return True
-        # (?<=X) or (?<!X) else (?<name>X)
-        return source[index + 3] in ["=", "!"]
 
-    @property
-    def group_builder(self):
-        return self._group_builder
+        # If it's a named group (e.g., (?P<name>...)), it's still a capturing group
+        if source[index + 2] == "P" and source[index + 3] == "<":
```

Comment:
Earlier we use a substring over a range and here we're using 2 diff substring char matches. 

In my head we should probably be using a range in all situations

Reply:
I'll take a look 👍